src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -869,35 +871,97 @@ internal static LLVMTypeRef GetLLVMTypeForTypeDesc(TypeDesc type)
                 case TypeFlags.ValueType:
                 case TypeFlags.Nullable:
                     {
-                        int structSize = type.GetElementSize().AsInt;
-
-                        // LLVM thinks certain sizes of struct have a different calling convention than Clang does.
-                        // Treating them as ints fixes that and is more efficient in general
-                        switch (structSize)
+                        if (!LlvmStructs.TryGetValue(type, out LLVMTypeRef llvmStructType))
                         {
-                            case 1:
-                                return LLVM.Int8Type();
-                            case 2:
-                                return LLVM.Int16Type();
-                            case 4:
-                                return LLVM.Int32Type();
-                            case 8:
-                                return LLVM.Int64Type();
-                        }
-
-                        int numInts = structSize / 4;
-                        int numBytes = structSize - numInts * 4;
-                        LLVMTypeRef[] structMembers = new LLVMTypeRef[numInts + numBytes];
-                        for (int i = 0; i < numInts; i++)
-                        {
-                            structMembers[i] = LLVM.Int32Type();
-                        }
-                        for (int i = 0; i < numBytes; i++)
-                        {
-                            structMembers[i + numInts] = LLVM.Int8Type();
+                            // LLVM thinks certain sizes of struct have a different calling convention than Clang does.
+                            // Treating them as ints fixes that and is more efficient in general
+                            int structSize = type.GetElementSize().AsInt;
+                            switch (structSize)
+                            {
+                                case 1:
+                                    llvmStructType = LLVM.Int8Type();
+                                    break;
+                                case 2:
+                                    llvmStructType = LLVM.Int16Type();
+                                    break;
+                                case 4:
+                                    llvmStructType = LLVM.Int32Type();
```

Comment:
It could cause interop issues, so I added logic to find structs that are really just float wrappers (or float wrapper wrappers). The trickier issue you made me realize is that a 4-byte struct made of two shorts just has an alignment of 2, so it should also be a full struct rather than an int32, so I've added alignment checks to all of these.